mlflow/utils/async_logging/async_artifacts_logging_queue.py

```diff
@@ -0,0 +1,263 @@
+        try:
+            run_artifacts = self._queue.get(timeout=1)
+        except Empty:
+            # Ignore empty queue exception
+            return
+
+        def logging_func(run_artifacts):
+            try:
+                self._logging_func(
+                    filename=run_artifacts.filename,
+                    artifact_path=run_artifacts.artifact_path,
+                    callback=run_artifacts.callback,
+                )
+
+                # Signal the batch processing is done.
+                run_artifacts.completion_event.set()
+
+            except Exception as e:
+                _logger.error(
+                    f"Run Id {run_artifacts.run_id}: Failed to log run data: Exception: {e}"
+                )
+                run_artifacts.exception = e
+                run_artifacts.completion_event.set()
+
+        self._batch_logging_worker_threadpool.submit(logging_func, run_artifacts)
+
+    def _wait_for_batch(self, artifacts: RunArtifacts) -> None:
+        """Wait for given artifacts to be processed by the logging thread.
+
+        Args:
+            artifacts: The artifacts to wait for.
+
+        Raises:
+            Exception: If an exception occurred while processing the batch.
+        """
+        artifacts.completion_event.wait()
+        if artifacts.exception:
+            raise artifacts.exception
+
+    def __getstate__(self):
+        """Return the state of the object for pickling.
+
+        This method is called by the `pickle` module when the object is being pickled. It returns a
+        dictionary containing the object's state, with non-picklable attributes removed.
+
+        Returns:
+            dict: A dictionary containing the object's state.
+        """
+        state = self.__dict__.copy()
+        del state["_queue"]
+        del state["_lock"]
+        del state["_is_activated"]
+
+        if "_run_data_logging_thread" in state:
+            del state["_run_data_logging_thread"]
+        if "_stop_data_logging_thread_event" in state:
+            del state["_stop_data_logging_thread_event"]
+        if "_batch_logging_thread" in state:
+            del state["_batch_logging_thread"]
+        if "_batch_logging_worker_threadpool" in state:
+            del state["_batch_logging_worker_threadpool"]
+        if "_batch_status_check_threadpool" in state:
+            del state["_batch_status_check_threadpool"]
+
+        return state
+
+    def __setstate__(self, state):
+        """Set the state of the object from a given state dictionary.
+
+        It pops back the removed non-picklable attributes from `self.__getstate__()`.
+
+        Args:
+            state (dict): A dictionary containing the state of the object.
+
+        Returns:
+            None
+        """
+        self.__dict__.update(state)
+        self._queue = Queue()
+        self._lock = threading.RLock()
+        self._is_activated = False
+        self._batch_logging_thread = None
```

Comment:
there is no `batch` here, let's call it `self._artifact_logging_thread`, and  the following as well.